### DISCUSSION TITLE
why the built output wheel file has permission 0600

### DISCUSSION THREAD

**Discussion by gaoweisgp** [apache/airflow#39769] [Q&A]
When I use `python -mbuild -w` to build the repo, the output wheel file has a permission of 0600, rather than 0644 as the hatch default sets?

```
% python -m build
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (GitPython==3.1.42, gitdb==4.0.11, hatchling==1.22.3, packaging==24.0, pathspec==0.12.1, pluggy==1.4.0, setuptools, smmap==5.0.1, tomli==2.0.1; python_version < '3.11', trove-classifiers==2024.3.3)
* Getting build dependencies for sdist...
* Building sdist...
* Building wheel from sdist
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (GitPython==3.1.42, gitdb==4.0.11, hatchling==1.22.3, packaging==24.0, pathspec==0.12.1, pluggy==1.4.0, setuptools, smmap==5.0.1, tomli==2.0.1; python_version < '3.11', trove-classifiers==2024.3.3)
* Getting build dependencies for wheel...
* Building wheel...
Successfully built apache_airflow-2.8.4.post2.tar.gz and apache_airflow-2.8.4-py3-none-any.whl
(venv) xxx@mmp-pw-dev /mnt/dev/repos/airflow (mseautosvc_3682_airflow_pbnj_test_01)✗ % ls -rtl dist
total 24976
-rw------- 1 xxx xxx 12331504 May 22 11:32 apache_airflow-2.8.4.tar.gz
-rw------- 1 xxx xxx 13239661 May 22 11:32 apache_airflow-2.8.4-py3-none-any.whl
```

**Comment by Taragolis** [comment]
Could you provide source of truth which stand for that `hatch` always produce distribution files (wheel and/or source distribution) with `0644`?

Check on macOS, even if I build `hatch` from the repo it is produce distribution with `0600`

```console
❯ git clone https://github.com/pypa/hatch.git
Cloning into 'hatch'...
remote: Enumerating objects: 39707, done.
remote: Counting objects: 100% (5172/5172), done.
remote: Compressing objects: 100% (1268/1268), done.
remote: Total 39707 (delta 2432), reused 4989 (delta 2341), pack-reused 34535
Receiving objects: 100% (39707/39707), 30.55 MiB | 3.31 MiB/s, done.
Resolving deltas: 100% (22446/22446), done.

❯ cd hatch

❯ hatch build
───────────────────────────────────────────────────────── sdist ─────────────────────────────────────────────────────────
dist/hatch-1.11.1.dev7.tar.gz
───────────────────────────────────────────────────────── wheel ─────────────────────────────────────────────────────────
dist/hatch-1.11.1.dev7-py3-none-any.whl

❯ ls -lah dist
total 5.1M
drwxr-xr-x  4 taragolis  128 May 23 12:19 .
drwxr-xr-x 23 taragolis  736 May 23 12:18 ..
-rw-------  1 taragolis 122K May 23 12:19 hatch-1.11.1.dev7-py3-none-any.whl
-rw-------  1 taragolis 5.0M May 23 12:19 hatch-1.11.1.dev7.tar.gz
```

**Reply by gaoweisgp** [reply]
I see it from this hatch code
https://github.com/pypa/hatch/blob/master/backend/src/hatchling/builders/utils.py#L96-L110

**Reply by gaoweisgp** [reply]
OK, this looks like to be hatch issue: https://github.com/pypa/hatch/issues/1519, need to normalize the wheel and zip files. Just need to pin or pull in the fixed hatch when it's resolved there

**Reply by Taragolis** [reply]
> https://github.com/pypa/hatch/blob/master/backend/src/hatchling/builders/utils.py#L96-L110

I guess this one only related to files permissions **inside** of archive (wheel or sdist)

**Reply by Taragolis** [reply]
Just curious what the problem you try to solve?

**Reply by gaoweisgp** [reply]
It's an internal multi-app group build issue where we build airflow first, and then pick the whl file up as dependency for the other app build. But this 0600 caused the wheel file failed to be picked up with permission error.

**Reply by potiuk** [reply]
You can always change the permission as post-processing step.  It's likely way easier than to even  open issue to hatch for them to change it. The main reason why there is 0600 is because of reproducibility that Hatch implements - which means that produced files have the same binary content. This is a security issue so highly unlikely someone would change it.

Adding `chmod g+r *.whl` in whatever build system you use seems like order of magnitued less hassle for everyone involved.

**Reply by mattp-** [reply]
@potiuk actually I've been working with @ofek via https://github.com/pypa/hatch/pull/1542 to fix the wheel/sdist builds to use 644 . once we get this merged and a new hatchling is released, if we can get that bubbled down to airflow it would be greatly appreciated :+1:

**Reply by potiuk** [reply]
Sure  - if you think it's worth it, it still order of magnitude more work that can be replaced with a single line of code in the CI scripts - but well - up to you :)

**Reply by potiuk** [reply]
There is nothing to do in Airflow - if new hatch will produce it this way, we will use it. The important thing for us is that permissions of all the produced files inside the packages do not change and produce stable, reproducible builds.

**Reply by mattp-** [reply]
hehe, unfortunately for $reasons at our work that aren't important or interesting we can't actually influence things like file perms, it just has to take what python -mbuild cranks out. will update here if/when we get a new hatchling release that airflow can bump. thanks :+1:

**Reply by potiuk** [reply]
Sure. FYI also work with @ofek on another smalll thing there - https://github.com/pypa/hatch/pull/1511 so I am paying attention :)
